### PR TITLE
Implement vec![expr; expr] and avoid unnecessary indent in chain

### LINKED ISSUE #1420: Chain with multi-line parent has bad indenting of first child
E.g.,

```
    given(
        r#"
        # Getting started
        ...
    "#,
    )
            .running(waltz)
```

I would expect `.running` to be indented once deeper than `given`. However, even that looks odd because the previous line is shorter than the tab width.

### LINKED ISSUE #1563: Extra indentation added when using `fn_call_style="Block"` and question mark operator
Consider this snippet:

```rust
fn query(conn: &Connection) -> Result<()> {
    conn.query_row(
        r#"
            SELECT title, date
            FROM posts,
            WHERE DATE(date) = $1
        "#,
        &[],
        |row| {
            Post {
                title: row.get(0),
                date: row.get(1),
            }
        },
    );

    Ok(())
}
```

When formatted with a question mark operator at the end, I get

```rust
fn query(conn: &Connection) -> Result<()> {
    conn.query_row(
            r#"
            SELECT title, date
            FROM posts,
            WHERE DATE(date) = $1
        "#,
            &[],
            |row| {
                Post {
                    title: row.get(0),
                    date: row.get(1),
                }
            },
        )?;

    Ok(())
}
```

I would expect:
```rust
fn query(conn: &Connection) -> Result<()> {
    conn.query_row(
        r#"
            SELECT title, date
            FROM posts,
            WHERE DATE(date) = $1
        "#,
        &[],
        |row| {
            Post {
                title: row.get(0),
                date: row.get(1),
            }
        },
    )?;

    Ok(())
}
```

### PR DESCRIPTION
**Pull request by topecongiro**
This PR closes #1563 and closes #1420.

### CODE REVIEW

**Comment by topecongiro** [comment]
Does a single argument with multiline needs trailing comma when using `fn_call_style = "Block"`?
I found that it gets removed after running rustfmt.
```rust
fn issue1420() {
    given(
        r#"
        # Getting started
        ...
    "#
    )
        .running(waltz)
}
```

**Comment by topecongiro** [comment]
Added commits to avoid weird formatting of macro in rhs of assignment.
Currently, rustfmt does not propagate the failure of rewriting macro([here](https://github.com/rust-lang-nursery/rustfmt/blob/master/src/expr.rs#L174-L175)). 
However, it should be acceptable to propagate the failure wherever we have a fallback path, like in `rewrite_assign_rhs`.

**Comment by nrc** [comment]
> Does a single argument with multiline needs trailing comma when using fn_call_style = "Block"?

Yes, e.g.,

```
foo(
    an_arg,
);
```

**Comment by nrc** [comment]
Unless it is a macro, in which case don't change the trailing comma-ness

File: tests/source/macros.rs [hunk 1]
```diff
@@ -34,6 +34,13 @@ fn main() {
     a,   
     ];
     
+    vec![a; b];
+    vec!(a; b);
+    vec!{a; b};
+
+    vec![a, b; c];
+    vec![a; b, c];
```
Comment (nrc) [inline tests/source/macros.rs]: Could you add some tests where the expression after the `;` is more complex and then gets reformatted?

**Comment by topecongiro** [comment]
Thank you!

File: tests/source/macros.rs (hunk 1 above)
Comment (topecongiro) [inline tests/source/macros.rs]: I added a few tests. Could you take a look if they suffices or there should be more tests?

Comment (nrc) [inline tests/source/macros.rs]: lgtm, thanks!

**Pull request merged**